services/site/src/middleware.ts

```diff
@@ -1,10 +1,73 @@
+import { from } from "@apollo/client";
+import { CurrentUserDocument, CurrentUserQuery } from "app/generated/graphql";
+import { createApolloClientRSC } from "app/lib/apollo-client/create-apollo-client-rsc";
+import { Cookie, createForwardCookiesToClientLink } from "app/lib/apollo-client/create-forward-cookies-to-client-link";
+import { SetCookieFunction } from "app/lib/apollo-client/create-forward-cookies-to-client-link";
+import { GetCookieHeaderFunction } from "app/lib/apollo-client/create-forward-cookies-to-server-link";
 import { NextRequest, NextResponse } from "next/server";
 
-export default function middleware(req: NextRequest): NextResponse {
-  const url = req.nextUrl.clone();
+import { getConfig } from "./lib/config/rsc";
 
-  if (process.env.SITE_GRAPHQL_ENDPOINT_SERVER && url.pathname === "/graphql") {
-    console.log(process.env.SITE_GRAPHQL_ENDPOINT_SERVER);
-    return NextResponse.rewrite(process.env.SITE_GRAPHQL_ENDPOINT_SERVER);
+type Middleware = {
+  match: (req: NextRequest) => boolean;
+  run: (req: NextRequest) => Promise<NextResponse | null>;
+};
+
+export default async function middleware(req: NextRequest): Promise<NextResponse | null> {
+  const middlewares = [forwardGraphqlRequests, redirectChallengeOverlayUrls, authentication];
+
+  for (const middleware of middlewares) {
```

Comment:
NextJS supports only a single middleware so we have to create our own middleware concept within that and check which functionality we actually want to execute.
We either forward the request to the gql server, redirect old challenge modal urls, or handle the authentication cookies